### ISSUE TITLE
Feature Request: Streaming Completion Requests

### ISSUE THREAD

**Issue by LLukas22** [EricLBuehler/mistral.rs#39]
It would be beneficial to have the capability to stream tokens to a client as they are generated, similar to other text generation interfaces. From what I've observed, this feature does not seem to be supported in the current `mistralrs-server` implementation.

Having briefly examined the `Engine` and `Sequence` implementations, it appears that sequences are indeed generated token by token, possibly in parallel, depending on the scheduler's discretion. With this process in mind, is it possible to introduce a signaling mechanism? This mechanism could be triggered after the `add_token` method of the `Sequence` is called, enabling the latest token to be sent back to the request processing thread. This would allow for the streaming of generated tokens back to the client in "real-time".

**Comment by EricLBuehler** [comment]
This seems like a great idea and has been something on my radar for a while. Like you pointed out, we currently only send responses from the `Engine` when a sequence (really a `SequenceGroup`) is done, and implementing some logic to handle streaming sequence groups would probably be the best way to do this. I will likely begin work on this in the next few days, but if you want to start, please feel free.

**Comment by EricLBuehler** [comment]
This was implemented in #57. Please feel free to reopen if you have any thoughts!

**Comment by LLukas22** [comment]
I'm impressed with how quickly you got this implemented! Today, I deployed some Mistral endpoints, and it appears that the streaming response is stripping tokens in the delta of the responses, removing all spaces and newlines.

The following script: 
```python
import openai

ENDPOINTS = {
    "MISTRALRS": "http://isonbgsrv32:12005/v1",
    "OLLAMA": "http://isonbgsrv32:11434/v1",
    "TGI": "http://isonbgsrv32:12001/v1"
}

for key in ENDPOINTS:
    client = openai.Client(api_key="foobar", base_url=ENDPOINTS[key])

    print(f"\n\nEndpoint: {key}")
    print("="*15)
    
    print("\nCompletion: ")
    print("="*15)
    answer = client.chat.completions.create(
        model="mistral",
        messages=[{"role": "user", "content": "What is the meaning of life?"}],
        stream=False,
        max_tokens=200,
        temperature=0.0 if key != "TGI" else 0.1
    )
    print(answer.choices[0].message.content)
       
    print("\n Streaming:")
    print("="*15)
    stream = client.chat.completions.create(
        model="mistral",
        messages=[{"role": "user", "content": "What is the meaning of life?"}],
        stream=True,
        max_tokens=200,
        temperature=0.1
    )

    for message in stream:
        print(message.choices[0].delta.content,flush=True, end="")
```

Produces the following output: 

```
Endpoint: MISTRALRS
===============

Completion:
===============
The meaning of life is a philosophical and personal question that has been debated throughout history. Different people may have different answers to this question, depending on their beliefs, values, and experiences. Some people believe that the meaning of life is to seek happiness, while others believe that it is to fulfill a specific purpose or to achieve a certain goal. Ultimately, the meaning of life is a deeply personal and subjective concept that each individual must define for themselves.</s>

 Streaming:
===============
Themeaningoflifeisaphilosophicalandpersonalquestionthathasbeendebatedthroughouthistory.Differentpeoplemayhavedifferentanswerstothisquestion,asitultimatelydependsonindividualbeliefs,values,andexperiences.Somemayfindmeaninginreligiousorspiritualbeliefs,whileothersmayfindmeaninginpursuingpersonalgoals,helpingothers,orcreatingmeaningfulrelationships.Ultimately,themeaningoflifeisadeeplypersonalandsubjectiveconceptthateachindividualmustexploreanddefineforthemselves.</s>

Endpoint: OLLAMA
===============

Completion:
===============
 The question of the meaning of life is a profound and complex philosophical issue that has puzzled humans for centuries. There are many different perspectives on this question, and no definitive answer can be given as it ultimately depends on one's personal beliefs and values. Some people believe that the meaning of life is to seek happiness and fulfillment, while others believe that it is to serve a higher power or to contribute to the betterment of humanity. Ultimately, each person must find their own answer to this question based on their individual beliefs and experiences.

 Streaming:
===============
 The question of the meaning of life is a profound and complex philosophical problem that has puzzled humans for centuries. There are many different perspectives on this question, and no definitive answer can be given as it ultimately depends on one's personal beliefs and values. Some people believe that the meaning of life is to seek happiness, knowledge, or personal growth, while others may see it as serving a higher power or purpose, or contributing to the greater good of humanity. Ultimately, each person must find their own answer to this question based on their individual beliefs and experiences.

Endpoint: TGI
===============

Completion:
===============
 The meaning of life is a question that has puzzled philosophers, theologians, and thinkers throughout history. There are many different perspectives and beliefs about what the meaning of life is. Some people believe that the meaning of life is to seek happiness, knowledge, or personal growth. Others believe that the meaning of life is to serve a higher power or to fulfill a specific purpose. Still others believe that life has no inherent meaning and that it is up to each individual to create their own meaning. Ultimately, the meaning of life is a deeply personal question and the answer may be different for each person. Some people may find comfort and guidance in religious or spiritual beliefs, while others may find meaning through their relationships, work, or creative pursuits. Ultimately, the search for the meaning of life is a lifelong journey and one that is unique to each individual.

 Streaming:
===============
 The meaning of life is a question that has puzzled philosophers, theologians, and thinkers throughout history. Many different beliefs and theories have been proposed, and ultimately, the answer to this question may depend on one's personal beliefs and values. Some people believe that the meaning of life is to seek happiness, knowledge, or personal fulfillment, while others believe that it is to serve a higher power or to contribute to the greater good of humanity. Still others believe that life has no inherent meaning, and that it is up to each individual to create their own purpose. Ultimately, the meaning of life may be a deeply personal and subjective question, and the answer may be different for each person.</s>

```
